[FSDP] Fix local_state_dict and state_dict_type bugs

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77101

1. state_dict_type does not correctly clean up self._state_dict.type.
2. local_state_dict has wrong assumptions that self.module.flat_param must present.

Differential Revision: [D36260234](https://our.internmc.facebook.com/intern/diff/D36260234/)

fegin committed:
[FSDP] Fix local_state_dict and state_dict_type bugs

1. state_dict_type does not correctly clean up self._state_dict.type.
2. local_state_dict has wrong assumptions that self.module.flat_param must present.

Differential Revision: [D36260234](https://our.internmc.facebook.com/intern/diff/D36260234/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, thanks for the quick fix!

zhaojuanmao reviewed (approved):
thanks for fixing those!

fegin committed:
Update on "[FSDP] Fix local_state_dict and state_dict_type bugs"


1. state_dict_type does not correctly clean up self._state_dict.type.
2. local_state_dict has wrong assumptions that self.module.flat_param must present.

Differential Revision: [D36260234](https://our.internmc.facebook.com/intern/diff/D36260234/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Fix local_state_dict and state_dict_type bugs"


1. state_dict_type does not correctly clean up self._state_dict.type.
2. local_state_dict has wrong assumptions that self.module.flat_param must present.

Differential Revision: [D36260234](https://our.internmc.facebook.com/intern/diff/D36260234/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Fix local_state_dict and state_dict_type bugs"


1. state_dict_type does not correctly clean up self._state_dict.type.
2. local_state_dict has wrong assumptions that self.module.flat_param must present.

Differential Revision: [D36260234](https://our.internmc.facebook.com/intern/diff/D36260234/)

[ghstack-poisoned]
